### PR TITLE
Adding Southpad Keyboard and default keymaps

### PR DESCRIPTION
**Pull request by MatchstickWorks**
## Description

Adding a new keyboard and default layout to the QMK code.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/matchstickworks/southpad/config.h [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6060
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    MatchstickWorks
+#define PRODUCT         SouthPad
+#define DESCRIPTION     Numpad
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/config.h:28]:
```suggestion
```

File: keyboards/matchstickworks/southpad/config.h [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+//#define MAGIC_KEY_LAYER7         7
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     B
+//#define MAGIC_KEY_BOOTLOADER_ALT ESC
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_EEPROM_CLEAR   BSPACE
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/config.h:172]:
```suggestion
```

File: keyboards/matchstickworks/southpad/config.h [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+//   #define RGBLIGHT_EFFECT_BREATHING
+//   #define RGBLIGHT_EFFECT_RAINBOW_MOOD
+//   #define RGBLIGHT_EFFECT_RAINBOW_SWIRL
+//   #define RGBLIGHT_EFFECT_SNAKE
+//   #define RGBLIGHT_EFFECT_KNIGHT
+//   #define RGBLIGHT_EFFECT_CHRISTMAS
+//   #define RGBLIGHT_EFFECT_STATIC_GRADIENT
+//   #define RGBLIGHT_EFFECT_RGB_TEST
+//   #define RGBLIGHT_EFFECT_ALTERNATING
+// #endif
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/config.h:80]:
```suggestion
```

File: keyboards/matchstickworks/southpad/config.h [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+/* Mechanical locking support. Use KC_LCAP, KC_LNUM or KC_LSCR instead in keymap */
+#define LOCKING_SUPPORT_ENABLE
+/* Locking resynchronize hack */
+#define LOCKING_RESYNC_ENABLE
+
+/* If defined, GRAVE_ESC will always act as ESC when CTRL is held.
+ * This is userful for the Windows task manager shortcut (ctrl+shift+esc).
+ */
+// #define GRAVE_ESC_CTRL_OVERRIDE
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/config.h:98]:
```suggestion
```

File: keyboards/matchstickworks/southpad/keymaps/default/config.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/* Copyright 2021 Logan Butler
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/keymaps/default/config.h:1]: This file should be removed.

File: keyboards/matchstickworks/southpad/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+}
+
+void matrix_scan_user(void) {
+
+}
+
+void led_set_user(uint8_t usb_led) {
+
+}
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/keymaps/default/keymap.c:40]:
```suggestion
const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
    [0] = LAYOUT_numpad_6x4_southpaw( /* Base */
        KC_BSPC, KC_EQL,  KC_LPRN, KC_RPRN, 
        KC_PMNS, KC_PAST, KC_PSLS, KC_NLCK, 
                 KC_P9,   KC_P8,   KC_P7, 
        KC_PPLS, KC_P6,   KC_P5,   KC_P4, 
                 KC_P3,   KC_P2,   KC_P1, 
        KC_PENT, KC_PDOT,    KC_P0
    ),
};
```

File: keyboards/matchstickworks/southpad/rules.mk [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+#     input clock frequency (before any prescaling is performed) in Hz. This value may
+#     differ from F_CPU if prescaling is used on the latter, and is required as the
+#     raw input clock is fed directly to the PLL sections of the AVR for high speed
+#     clock generation for the USB and other AVR subsections. Do NOT tack on a 'UL'
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/rules.mk:38]:
```suggestion
# Bootloader selection
BOOTLOADER = bootloadHID

F_CPU = 16000000
```

File: keyboards/matchstickworks/southpad/rules.mk [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/rules.mk:59]:
```suggestion
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
UNICODE_ENABLE = no         # Unicode
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/matchstickworks/southpad/southpad.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	// runs for every action, just before processing by the firmware
+
+	return process_record_user(keycode, record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+	// put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+	led_set_user(usb_led);
+}
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/southpad.c:43]:
```suggestion
```

File: keyboards/matchstickworks/southpad/southpad.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+ * represents the switch matrix.
+ */
+#define LAYOUT_numpad_6x4( \
+               k00, k01, k02, k03,      \
+               k10, k11, k12, k13,      \
+               k20, k21, k22, k23,      \
+					k31, k32, k33,      \
+               k40, k41, k42, k43,      \
+					k51, 	  k53       \
+) \
```
Comment (drashna) [inline keyboards/matchstickworks/southpad/southpad.h:35]:
```suggestion
#define LAYOUT_numpad_6x4_southpaw( \
               k00, k01, k02, k03,  \
               k10, k11, k12, k13,  \
                    k21, k22, k23,  \
               k20, k31, k32, k33,  \
                    k41, k42, k43,  \
               k40, k51,   k53      \
) \
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/matchstickworks/southpad/keymaps/default/keymap.c (hunk 6 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/keymaps/default/keymap.c]: Committing the "_southpaw" suggestion as well as the rest of the cleanup. However, PCB is not wired this way, this mapping does not work.

File: keyboards/matchstickworks/southpad/southpad.h (hunk 10 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/southpad.h]: Committed formatting cleanup as well as naming, remapping of layout does not work (similar to keymap.c suggestion).

File: keyboards/matchstickworks/southpad/keymaps/default/config.h (hunk 5 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/keymaps/default/config.h:1]: File to be removed in next merge

File: keyboards/matchstickworks/southpad/southpad.h (hunk 10 above)
Comment (fauxpark) [reply keyboards/matchstickworks/southpad/southpad.h]: It does work if you apply both of these suggestions. The electrical mappings haven't been changed.

Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/southpad.h]: Roger, I will try it again

Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/southpad.h]: Ah, it was my poor assumption of changing the mapping in the curly brackets of southpad.h
Thanks for the help!

File: keyboards/matchstickworks/southpad/keymaps/default/keymap.c (hunk 6 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/keymaps/default/keymap.c]: Resolved with comments on southpad.h

**Review by drashna — APPROVED** [review]

**Comment by MatchstickWorks** [comment]
Hey @drashna It looks like the renaming of my layout has failed the lint check. Can you enlighten me on what the issue is here?

File: keyboards/matchstickworks/southpad/rules.mk [hunk 11]
```diff
@@ -0,0 +1,28 @@
+# MCU name
+MCU = atmega32u2
+
+# Bootloader selection
+BOOTLOADER = bootloadHID
+
+F_CPU = 16000000
+
```
Comment (fauxpark) [inline keyboards/matchstickworks/southpad/rules.mk:8]:
```suggestion
# Processor frequency
F_CPU = 16000000

# Bootloader selection
BOOTLOADER = bootloadHID
```

File: keyboards/matchstickworks/southpad/rules.mk [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+LAYOUTS = numpad_6x4_southpaw
```
Comment (fauxpark) [inline keyboards/matchstickworks/southpad/rules.mk:28]:
```suggestion
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Also missing an info.json.

**Comment by MatchstickWorks** [comment]
> Also missing an info.json.

Hey @fauxpark, thanks for the review! Do you mind pointing me to where I can learn more about the info.json file? Or at the very least a solid keyboard I can base my own file off of? I'm seeing a few different guides but none of them seem to be playing nice with the QMK Configurator tool.

**Comment by drashna** [comment]
If you have a json from kbfirmware, then https://noroadsleft.github.io/kbf_qmk_converter/ can do the heavy lifting.

File: keyboards/matchstickworks/southpad/info.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+                {"label":"K50 (C6,B0)", "x":0, "y":4, "h":2},
+                {"label":"K41 (C7,B1)", "x":1, "y":4},
+                {"label":"K42 (C7,B2)", "x":2, "y":4},
+                {"label":"K43 (C7,B4)", "x":3, "y":4},
+                {"label":"K51 (C6,B1)", "x":1, "y":5},
+                {"label":"K53 (C6,B4)", "x":2, "y":5, "w":2}
+            ]
+        }
+    }
+    ,"meta": "https://noroadsleft.github.io/kbf_qmk_converter/"
```
Comment (fauxpark) [inline keyboards/matchstickworks/southpad/info.json:34]:
```suggestion
```

File: keyboards/matchstickworks/southpad/readme.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
+# SouthPad
+
+![Southpad](https://i.imgur.com/Wl71thn.jpg)
```
Comment (fauxpark) [inline keyboards/matchstickworks/southpad/readme.md:3]:
```suggestion
![Southpad](https://i.imgur.com/Wl71thnh.jpg)
```

File: keyboards/matchstickworks/southpad/rules.mk [hunk 15]
```diff
@@ -0,0 +1,25 @@
+# MCU name
+MCU = atmega32u2
+
+# Processor frequency
+F_CPU = 16000000
+
```
Comment (fauxpark) [inline keyboards/matchstickworks/southpad/rules.mk:6]:
```suggestion
```
Whoops, this actually isn't needed at all; the default frequency for the AVRs (with one exception) is 16MHz.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/matchstickworks/southpad/readme.md (hunk 14 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/readme.md:3]: Presuming the appended "h" limits the size of the image? TIL!

File: keyboards/matchstickworks/southpad/rules.mk (hunk 15 above)
Comment (MatchstickWorks) [reply keyboards/matchstickworks/southpad/rules.mk:6]: Ah! Wasn't sure if QMK knew to make that assumption. Good to know for the future!

**Review by fauxpark — APPROVED** [review]

**Pull request merged**